### PR TITLE
fix: [PL-31037]: Added missing setup usage for secrets used in ssh and winrm secrets

### PR DESCRIPTION
**Pull request by nishantsaini20**
## Describe your changes
SSH and Win RM secret are special kind of secrets which also can refer another secret. Setup usage is secrets referred by ssh/winrm secret was missing. As part of this PR creation and deletion of setup usage for a secret referencing another secret is introduced.

Testing:
1. On creation of SSH/WinRm secret any secret reference should create a setup.
Screenshot 2023-03-18 at 5 06 36 PM
2. If a secret is no longer reference the corresponding setup usage should be deleted.
Screenshot 2023-03-18 at 5 08 43 PM
3. If the SSH secret is deleted any setup usage of any secret referred by ssh secret should get deleted.
Screenshot 2023-03-18 at 5 10 37 PM
4. Also tested update flow by changing the secret being referred. Reference again older secret gets removed and gets added for changed secret as expected.
5. Same as above for WinRm secret as well.

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger utAll`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/45182)

### CODE REVIEW

**Comment by nishantsaini20** [comment]
trigger ut8

**Comment by nishantsaini20** [comment]
trigger ut6

**Review by gn-harness — APPROVED** [review]
> Reviewed 26 of 26 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45182)* status: 26 file(s) to review. 1 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/api/impl/SecretEntityReferenceHelper.java` line 122 at r1](https://reviewable.io/reviews/harness/harness-core/45182#-NR0P45-Al7LeAnUmGHX:-NR0P45-Al7LeAnUmGHY:b-p4xzo8) ([raw file](https://github.com/harness/harness-core/blob/882d6b5c5544bf10f28aa4e939548afafb18069d/120-ng-manager/src/main/java/io/harness/ng/core/api/impl/SecretEntityReferenceHelper.java#L122)):*
> ```Java
>   }
> 
>   public void createSetupUsageForSecret(String accountIdentifier, SecretDTOV2 secretDTOV2) {
> ```

nit: can you please add some documentation 



_Code quote:_
```Java
public void createSetupUs
```
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/api/impl/SecretEntityReferenceHelper.java` line 158 at r1](https://reviewable.io/reviews/harness/harness-core/45182#-NR0DLk_CfyBMLoiFsVw:-NR0DLk_CfyBMLoiFsVx:b-1r97uw) ([raw file](https://github.com/harness/harness-core/blob/882d6b5c5544bf10f28aa4e939548afafb18069d/120-ng-manager/src/main/java/io/harness/ng/core/api/impl/SecretEntityReferenceHelper.java#L158)):*
> ```Java
>       log.info(ENTITY_REFERENCE_LOG_PREFIX
>               + "Entity reference was not updated for the secret [{}] using the secret [{}] with the exception[{}]",
>           secretFQN, referredSecretFQNs, ex.getMessage());
> ```

nit: preserve the stack trace too



_Code quote:_
```Java
ex.getMessage())
```
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/api/impl/SecretEntityReferenceHelper.java` line 184 at r1](https://reviewable.io/reviews/harness/harness-core/45182#-NR0FW5h1T1Oyu23ozTX:-NR0FW5h1T1Oyu23ozTY:b-j8sa3) ([raw file](https://github.com/harness/harness-core/blob/882d6b5c5544bf10f28aa4e939548afafb18069d/120-ng-manager/src/main/java/io/harness/ng/core/api/impl/SecretEntityReferenceHelper.java#L184)):*
> ```Java
>     List<EntityDetailWithSetupUsageDetailProtoDTO> allSecretDetails = new ArrayList<>();
>     for (Map.Entry<String, SecretRefData> secret : secrets.entrySet()) {
>       if (secret != null && secret.getValue() != null && !secret.getValue().isNull()) {
> ```

unnest

if (secret == null || secret.getValue() == null && secret.getValue().isNull()) {
  continue
}
___
*[`950-delegate-tasks-beans/src/main/java/io/harness/ng/core/dto/secrets/SecretSpecDTO.java` line 38 at r1](https://reviewable.io/reviews/harness/harness-core/45182#-NR0RMV99Nq_Z2xPDUCv:-NR0RMV99Nq_Z2xPDUCw:b-86xd2e) ([raw file](https://github.com/harness/harness-core/blob/882d6b5c5544bf10f28aa4e939548afafb18069d/950-delegate-tasks-beans/src/main/java/io/harness/ng/core/dto/secrets/SecretSpecDTO.java#L38)):*
> ```Java
>   public Optional<List<DecryptableEntity>> getDecryptableEntities() {
>     return Optional.empty();
>   }
> ```

why both Optional and List?



_Code quote:_
```Java
  public Optional<List<DecryptableEntity>> getDecryptableEntities() {
    return Optional.empty();
  }
```
___
*[`950-delegate-tasks-beans/src/main/java/io/harness/ng/core/dto/secrets/WinRmCredentialsSpecDTO.java` line 65 at r1](https://reviewable.io/reviews/harness/harness-core/45182#-NR0Sg7X9MfwRXG9CWAT:-NR0Sg7X9MfwRXG9CWAU:b-83eeib) ([raw file](https://github.com/harness/harness-core/blob/882d6b5c5544bf10f28aa4e939548afafb18069d/950-delegate-tasks-beans/src/main/java/io/harness/ng/core/dto/secrets/WinRmCredentialsSpecDTO.java#L65)):*
> ```Java
>     } else if (auth.getAuthScheme() == WinRmAuthScheme.Kerberos) {
>       KerberosWinRmConfigDTO spec = (KerberosWinRmConfigDTO) auth.getSpec();
>       return Optional.ofNullable(spec.getSpec().getDecryptableEntities());
> ```

these can be null?



_Code quote:_
```Java
Optional.ofNullable
```

**Comment by meenakshi-harness** [comment]
can we also test if new secret is referenced in update call, setup usage should also change

**Comment by nishantsaini20** [comment]
trigger ut6

**Comment by nishantsaini20** [comment]
trigger messagecheck

File: 120-ng-manager/src/main/java/io/harness/ng/core/api/impl/SecretEntityReferenceHelper.java [hunk 1]
```diff
@@ -100,6 +119,98 @@ public void createSetupUsageForSecretManager(String accountIdentifier, String or
     }
   }
 
+  public void createSetupUsageForSecret(String accountIdentifier, SecretDTOV2 secretDTOV2) {
+    String secretFQN = FullyQualifiedIdentifierHelper.getFullyQualifiedIdentifier(accountIdentifier,
+        secretDTOV2.getOrgIdentifier(), secretDTOV2.getProjectIdentifier(), secretDTOV2.getIdentifier());
+    Optional<List<DecryptableEntity>> decryptableEntities = secretDTOV2.getSpec().getDecryptableEntities();
+    if (decryptableEntities.isEmpty()) {
+      produceSetupUsageForSecretEvent(accountIdentifier, secretDTOV2, secretFQN, new ArrayList<>(), new ArrayList<>());
```
[the remark this answers is not in the document]
Comment (nishantsaini20) [reply 120-ng-manager/src/main/java/io/harness/ng/core/api/impl/SecretEntityReferenceHelper.java:133]: User can switch auth mechanism of ssh to `Username/SSH Key File` Path from any other auth mechanism.  `Username/SSH Key File` can have zero secret references. In such a case the older references should be deleted. This case handles it.

**Review by nishantsaini20 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45182)* status: 26 file(s) to review. 5. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/api/impl/SecretEntityReferenceHelper.java` line 158 at r1](https://reviewable.io/reviews/harness/harness-core/45182#-NR0DLk_CfyBMLoiFsVw:-NSWh-bc7OWebrdxmu6Y:b-1w4wfe) ([raw file](https://github.com/harness/harness-core/blob/882d6b5c5544bf10f28aa4e939548afafb18069d/120-ng-manager/src/main/java/io/harness/ng/core/api/impl/SecretEntityReferenceHelper.java#L158)):*
<details><summary><i>Previously, gn-harness (Gaurav Nanda) wrote…</i></summary><blockquote>

nit: preserve the stack trace too
</blockquote></details>

exception stack trace will get logged by the next line.
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/api/impl/SecretEntityReferenceHelper.java` line 184 at r1](https://reviewable.io/reviews/harness/harness-core/45182#-NR0FW5h1T1Oyu23ozTX:-NSWhKFVEr_djGbXkaRq:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/882d6b5c5544bf10f28aa4e939548afafb18069d/120-ng-manager/src/main/java/io/harness/ng/core/api/impl/SecretEntityReferenceHelper.java#L184)):*
<details><summary><i>Previously, gn-harness (Gaurav Nanda) wrote…</i></summary><blockquote>

unnest

if (secret == null || secret.getValue() == null && secret.getValue().isNull()) {
  continue
}
</blockquote></details>

Done.
___
*[`950-delegate-tasks-beans/src/main/java/io/harness/ng/core/dto/secrets/SecretSpecDTO.java` line 38 at r1](https://reviewable.io/reviews/harness/harness-core/45182#-NR0RMV99Nq_Z2xPDUCv:-NRZ-p8R-L9n00p7lSke:bbjg3ty) ([raw file](https://github.com/harness/harness-core/blob/882d6b5c5544bf10f28aa4e939548afafb18069d/950-delegate-tasks-beans/src/main/java/io/harness/ng/core/dto/secrets/SecretSpecDTO.java#L38)):*
<details><summary><i>Previously, gn-harness (Gaurav Nanda) wrote…</i></summary><blockquote>

why both Optional and List?
</blockquote></details>

There is a possibility of nullable value in one of the deep nesting. Kept it for safe case.
___
*[`950-delegate-tasks-beans/src/main/java/io/harness/ng/core/dto/secrets/WinRmCredentialsSpecDTO.java` line 65 at r1](https://reviewable.io/reviews/harness/harness-core/45182#-NR0Sg7X9MfwRXG9CWAT:-NRYxDjdDm4TBMDNumOy:b-3ss1ii) ([raw file](https://github.com/harness/harness-core/blob/882d6b5c5544bf10f28aa4e939548afafb18069d/950-delegate-tasks-beans/src/main/java/io/harness/ng/core/dto/secrets/WinRmCredentialsSpecDTO.java#L65)):*
<details><summary><i>Previously, gn-harness (Gaurav Nanda) wrote…</i></summary><blockquote>

these can be null?
</blockquote></details>

Yes there is possibility of null value.

**Review by gn-harness — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45182)* status: 26 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by meenakshi-harness — APPROVED** [review]
:lgtm:

> Reviewed 11 of 26 files at r1, 15 of 15 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45182#-:-NSdjfLg07JY3Rg7LRZg:bnfp4nl)* status: 15 file(s) to review. 1. unresolved discussions.

**Pull request merged**